File: public/agentsmem_tool.py
#!/usr/bin/env python3
"""
AgentsMem memory encrypt/decrypt tool (Python).
Usage: --gen-key | --encrypt | --decrypt with --key, --in, --out; optional --md5 for verify on decrypt.
"""
import os
import hashlib
import argparse
import secrets

def calculate_md5(file_path):
    """Compute MD5 of file."""
    h = hashlib.md5()
    with open(file_path, "rb") as f:
        for chunk in iter(lambda: f.read(4096), b""):
            h.update(chunk)
    return h.hexdigest()

def get_keystream(key_str, salt, length):
    seed = hashlib.pbkdf2_hmac("sha256", key_str.encode(), salt, 100000, 32)
    stream = bytearray()
    counter = 0
    while len(stream) < length:
        block = hashlib.sha256(seed + counter.to_bytes(4, "big")).digest()
        stream.extend(block)
        counter += 1
    return stream[:length]

def encrypt_file(key_str, in_p, out_p):
    if not os.path.isfile(in_p):
        print("Error: source file not found")
        return
    salt = os.urandom(16)
    with open(in_p, "rb") as f:
        data = f.read()
    keystream = get_keystream(key_str, salt, len(data))
    res = bytes([b ^ k for b, k in zip(data, keystream)])
    with open(out_p, "wb") as f:
        f.write(salt + res)
    md5_val = calculate_md5(out_p)
    print(f"Encrypt OK: {out_p}")
    print(f"Ciphertext MD5: {md5_val} (use this for x-ciphertext-md5 when uploading)")

def decrypt_file(key_str, in_p, out_p, expected_md5=None):
    if not os.path.isfile(in_p):
        print("Error: encrypted file not found")
        return
    if expected_md5:
        actual_md5 = calculate_md5(in_p)
        if actual_md5.lower() != expected_md5.lower():
            print(f"MD5 mismatch; file may be corrupted. Expected: {expected_md5} Actual: {actual_md5}")
            return
        print("MD5 OK, decrypting...")
    with open(in_p, "rb") as f:
        data = f.read()
    salt, payload = data[:16], data[16:]
    keystream = get_keystream(key_str, salt, len(payload))
    res = bytes([b ^ k for b, k in zip(payload, keystream)])
    with open(out_p, "wb") as f:
        f.write(res)
    print(f"Decrypt OK: {out_p}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="AgentsMem encrypt/decrypt tool (Python)")
    parser.add_argument("--gen-key", action="store_true", help="Generate a new encryption key")
    parser.add_argument("--encrypt", action="store_true")
    parser.add_argument("--decrypt", action="store_true")
    parser.add_argument("--key", type=str, help="Encryption key")
    parser.add_argument("--in", dest="in_p", metavar="FILE", help="Input file")
    parser.add_argument("--out", dest="out_p", metavar="FILE", help="Output file")
    parser.add_argument("--md5", dest="check_md5", type=str, help="Verify ciphertext MD5 before decrypt (optional)")
    args = parser.parse_args()

    if args.gen_key:
        print(f"Key: {secrets.token_urlsafe(32)}")
    elif args.encrypt and args.key and args.in_p and args.out_p:
        encrypt_file(args.key, args.in_p, args.out_p)
    elif args.decrypt and args.key and args.in_p and args.out_p:
        decrypt_file(args.key, args.in_p, args.out_p, args.check_md5)
    else:
        parser.print_help()

File: public/agentsmem_tool.js
#!/usr/bin/env node
/**
 * AgentsMem memory encrypt/decrypt tool (Node.js).
 * Usage: --gen-key | --encrypt | --decrypt with --key, --in, --out; optional --md5 for verify on decrypt.
 */
const crypto = require("crypto");
const fs = require("fs");

function calculateMD5(filePath) {
  const buffer = fs.readFileSync(filePath);
  return crypto.createHash("md5").update(buffer).digest("hex");
}

function getKeystream(keyStr, salt, length) {
  const seed = crypto.pbkdf2Sync(keyStr, salt, 100000, 32, "sha256");
  let stream = Buffer.alloc(0);
  let counter = 0;
  while (stream.length < length) {
    const counterBuf = Buffer.alloc(4);
    counterBuf.writeUInt32BE(counter);
    const hash = crypto.createHash("sha256").update(Buffer.concat([seed, counterBuf])).digest();
    stream = Buffer.concat([stream, hash]);
    counter++;
  }
  return stream.slice(0, length);
}

function processFile(keyStr, inP, outP, isEncrypt, expectedMd5 = null) {
  try {
    if (!fs.existsSync(inP)) {
      console.error("Error: input file not found");
      return;
    }
    if (!isEncrypt && expectedMd5) {
      const actualMd5 = calculateMD5(inP);
      if (actualMd5.toLowerCase() !== expectedMd5.toLowerCase()) {
        console.error(`MD5 mismatch; file may be corrupted. Expected: ${expectedMd5} Actual: ${actualMd5}`);
        return;
      }
      console.log("MD5 OK, decrypting...");
    }
    const data = fs.readFileSync(inP);
    let salt, payload;
    if (isEncrypt) {
      salt = crypto.randomBytes(16);
      payload = data;
    } else {
      salt = data.slice(0, 16);
      payload = data.slice(16);
    }
    const keystream = getKeystream(keyStr, salt, payload.length);
    const res = Buffer.alloc(payload.length);
    for (let i = 0; i < payload.length; i++) res[i] = payload[i] ^ keystream[i];
    if (isEncrypt) {
      fs.writeFileSync(outP, Buffer.concat([salt, res]));
      const md5 = calculateMD5(outP);
      console.log(`Encrypt OK: ${outP}`);
      console.log(`Ciphertext MD5: ${md5} (use this for x-ciphertext-md5 when uploading)`);
    } else {
      fs.writeFileSync(outP, res);
      console.log(`Decrypt OK: ${outP}`);
    }
  } catch (e) {
    console.error(`Error: ${e.message}`);
  }
}

// Simple CLI parsing
const args = process.argv.slice(2);
const p = {};
for (let i = 0; i < args.length; i++) {
  if (args[i].startsWith("--")) p[args[i].replace("--", "")] = args[i + 1] || true;
}

if (p["gen-key"]) {
  console.log("Key:", crypto.randomBytes(32).toString("base64url"));
} else if (p.encrypt) {
  if (p.key && p.in && p.out) processFile(p.key, p.in, p.out, true);
  else console.log("Usage: node agentsmem_tool.js --encrypt --key KEY --in FILE --out FILE");
} else if (p.decrypt) {
  if (p.key && p.in && p.out) processFile(p.key, p.in, p.out, false, p.md5 || null);
  else console.log("Usage: node agentsmem_tool.js --decrypt --key KEY --in FILE --out FILE [--md5 HEX]");
} else {
  console.log("AgentsMem encrypt/decrypt tool.");
  console.log("Examples:");
  console.log("  node agentsmem_tool.js --gen-key");
  console.log("  node agentsmem_tool.js --encrypt --key K --in F1 --out F2");
  console.log("  node agentsmem_tool.js --decrypt --key K --in F2 --out F3 [--md5 M]");
}
